pkg/kube/object.go

```diff
@@ -84,3 +88,113 @@ func (ci ContainerImages) AsJSON() (string, error) {
 func (ci ContainerImages) FromJSON(value string) error {
 	return json.Unmarshal([]byte(value), &ci)
 }
+
+func GVKForObject(obj runtime.Object, scheme *runtime.Scheme) (schema.GroupVersionKind, error) {
+	gvks, isUnversioned, err := scheme.ObjectKinds(obj)
+	if err != nil {
+		return schema.GroupVersionKind{}, err
+	}
+	if isUnversioned {
+		return schema.GroupVersionKind{}, fmt.Errorf("cannot create group-version-kind for unversioned type %T", obj)
+	}
+
+	if len(gvks) < 1 {
+		return schema.GroupVersionKind{}, fmt.Errorf("no group-version-kinds associated with type %T", obj)
+	}
+	if len(gvks) > 1 {
+		// this should only trigger for things like metav1.XYZ --
+		// normal versioned types should be fine
+		return schema.GroupVersionKind{}, fmt.Errorf(
+			"multiple group-version-kinds associated with type %T, refusing to guess at one", obj)
+	}
+	return gvks[0], nil
+}
+
+func KindForObject(object metav1.Object, scheme *runtime.Scheme) (string, error) {
+	ro, ok := object.(runtime.Object)
+	if !ok {
+		return "", fmt.Errorf("%T is not a runtime.Object", object)
+	}
+	gvk, err := GVKForObject(ro, scheme)
+	if err != nil {
+		return "", err
+	}
+	return gvk.Kind, nil
+}
+
+func SetOwnerReference(owner, object metav1.Object, scheme *runtime.Scheme) error {
+	// Validate the owner.
+	ro, ok := owner.(runtime.Object)
+	if !ok {
+		return fmt.Errorf("%T is not a runtime.Object, cannot call SetOwnerReference", owner)
+	}
+	if err := validateOwner(owner, object); err != nil {
+		return err
+	}
+
+	// Create a new owner ref.
+	gvk, err := GVKForObject(ro, scheme)
+	if err != nil {
+		return err
+	}
+	ref := metav1.OwnerReference{
+		APIVersion: gvk.GroupVersion().String(),
+		Kind:       gvk.Kind,
+		UID:        owner.GetUID(),
+		Name:       owner.GetName(),
+	}
+
+	// Update owner references and return.
+	upsertOwnerRef(ref, object)
+	return nil
+}
+
+func validateOwner(owner, object metav1.Object) error {
+	ownerNs := owner.GetNamespace()
+	if ownerNs != "" {
+		objNs := object.GetNamespace()
+		if objNs == "" {
+			return fmt.Errorf("cluster-scoped resource must not have a namespace-scoped owner, owner's namespace %s", ownerNs)
+		}
+		if ownerNs != objNs {
+			return fmt.Errorf("cross-namespace owner references are disallowed, owner's namespace %s, obj's namespace %s", owner.GetNamespace(), object.GetNamespace())
+		}
+	}
+	return nil
+}
+
+func upsertOwnerRef(ref metav1.OwnerReference, object metav1.Object) {
+	owners := object.GetOwnerReferences()
+	idx := indexOwnerRef(owners, ref)
+	if idx == -1 {
+		owners = append(owners, ref)
+	} else {
+		owners[idx] = ref
```

Comment:
Just for my understanding:
In this case, what field will be updated? Only `UID` considering the unit test? I'm curious when it happens. It looks like `Group`, `Kind`, and `Name` are always the same.

Reply:
In this case all properties of the given `metav1.OwnerReference` are updated including GVK even though it's very unlikely they change in real Starboard scenario. However this utility method is generic to cater for cases where we have some other controller / Kubernetes app that also relies on the `ownerReferences` property. In this case the GVK might be different.

Reply:
If `indexOwnerRef` returns `idx` other than `-1`, it means GVK is the same, right? So, I wanted to know why `owners[idx] = ref` is necessary. If all properties are the same as existing `ownerReference`, we don't need to update it. But `UID` might be different when a user deletes the resource and recreates it with the same GVK and name.

Reply:
Yup. You're right. It might only happen if the UID has changed.

Reply:
I see. Thank you for the clarification!